background.ts

```diff
@@ -747,19 +771,248 @@ function newK8sManager() {
+      }
+    } catch (err) {
+      errors.push(`JSON-parsing error checking field ${ fqname }: ${ err }`);
+    }
+
+    return false;
+  }
+
+  /**
+   * Check semantics of SET commands:
+   * - verify that setting names are recognized, and validate provided values
+   * - returns an array of two strings:
+   *   1. a description of the status of the request, if it was valid
+   *   2. a list of any errors in the request body.
+   * @param newSettings: a subset of the Settings object, containing the desired values
+   * @returns [{string} description of final state if no error, {string} error message]
+   */
+  async updateSettings(newSettings: Record<string, any>): Promise<[string, string]> {
+    const allowedSettings: Record<string, validationFunc|any> = {
+      version:    this.verifyUnchanged,
+      kubernetes: {
+        version:                    this.verifyKubernetesVersion,
+        memoryInGB:                 this.verifyUnchanged,
+        numberCPUs:                 this.verifyUnchanged,
+        port:                       this.verifyUnchanged,
+        containerEngine:            this.verifyContainerEngine,
+        checkForExistingKimBuilder: this.verifyUnchanged,
+        enabled:                    this.verifyEnabled,
+        WSLIntegrations:            this.verifyObjectUnchanged,
+        options:                    { traefik: this.verifyUnchanged },
+      },
+      portForwarding: { includeKubernetesServices: this.verifyUnchanged },
+      images:         {
+        showAll:   this.verifyUnchanged,
+        namespace:  this.verifyUnchanged,
+      },
+      telemetry: this.verifyUnchanged,
+      updater:   this.verifyUnchanged,
+      debug:     this.verifyUnchanged
+    };
+    const errors: Array<string> = [];
+
+    if (this.k8sVersions.length === 0) {
+      this.k8sVersions = (await k8smanager.availableVersions).map(entry => entry.version.version);
+    }
+    const needToUpdate = this.verifyProposedSettings('', allowedSettings, newSettings, errors);
+
+    if (errors.length > 0) {
+      return ['', `errors in attempt to update settings:\n${ errors.join('\n') }`];
+    }
+    if (needToUpdate) {
+      writeSettings(newSettings);
+      // cfg is a global, and at this point newConfig has been merged into it :(
+      window.send('settings-update', cfg);
+      if (!BackendIsBusy()) {
+        pendingRestart = false;
+        setImmediate(doFullRestart);
+
+        return ['triggering a restart to apply changes', ''];
+      } else {
+        // Call doFullRestart once the UI is finished starting or stopping
+        pendingRestart = true;
+
+        return ['UI is currently busy, but will eventually restart to apply changes', ''];
+      }
+    } else {
+      return ['no changes necessary', ''];
+    }
+  }
+
+  async requestShutdown() {
+    await k8smanager.stop();
     Electron.app.quit();
   }
 }
+
+const httpCommandServer = new HttpCommandServer(new BackgroundCommandWorker());
```

Comment:
You can't refer to classes at the global level, at least in TypeScript. I was trying to avoid the `let instance: type|null = null` thing by assigning it at the global level.